Re-add ScheduledStartTime to orchestration scheduling in .NET isolated

davidmrdavid opened the pull request:
<!-- Start the PR description with some context for the change. -->
Follow up to:
https://github.com/Azure/azure-functions-durable-extension/pull/2634
https://github.com/Azure/azure-functions-durable-extension/pull/2544#issuecomment-1688579724

Fixes: https://github.com/microsoft/durabletask-dotnet/issues/256

When refactoring the localGrpcListener (used by .NET isolated and Java) to use an IDurableClient (necessary for Distributed Tracing), we accidentally dropped the ability to set an orchestrator's "fireAt" time. This PR adds that back.

Since the IDurableClient does not expose functionality for scheduled orchestrators, this PR constructs the `TaskHubClient`, which does expose that functionality. It also leaves a few "TODO" comments to simplify the implementation in the future.
 
 **Sitll pending:**
 * Manual E2E test
 * Inclusion of distributed tracing regression test
 * Inclusion of test of scheduled orchestrations test.
 
<!-- Make sure to delete the markdown comments and the below sections when squash merging -->
### Issue describing the changes in this PR

resolves https://github.com/microsoft/durabletask-dotnet/issues/256


### Pull request checklist

* [ ] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation PR is ready to merge and referenced in `pending_docs.md`
* [ ] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [ ] I have added all required tests (Unit tests, E2E tests)
* [ ] My changes **do not** require any extra work to be leveraged by OutOfProc SDKs
    * [ ] Otherwise: That work is being tracked here: #issue_or_pr_in_each_sdk
* [ ] My changes **do not** change the version of the WebJobs.Extensions.DurableTask package
    * [ ] Otherwise: major or minor version updates are reflected in `/src/Worker.Extensions.DurableTask/AssemblyInfo.cs`
* [ ] My changes **do not** add EventIds to our EventSource logs
    * [ ] Otherwise: Ensure the EventIds are within the supported range in our existing Windows infrastructure. You may validate this with a deployed app's telemetry. You may also extend the range by completing a PR such as [this one](https://msazure.visualstudio.com/One/_git/AAPT-Antares-Websites/pullrequest/7463263?_a=files).
* [ ] My changes **should** be added to v3.x branch.
    * [ ] Otherwise: This change only applies to Durable Functions v2.x and **will not** be merged to branch v3.x.

davidmrdavid committed:
naive first implementation

cgillum commented on src/WebJobs.Extensions.DurableTask/LocalGrpcListener.cs:
```diff
@@ -153,13 +153,33 @@ public override Task<P.DeleteTaskHubResponse> DeleteTaskHub(P.DeleteTaskHubReque
+                        await this.GetDurabilityProvider(context).CreateTaskOrchestrationAsync(
+                        new TaskMessage
+                        {
+                            Event = new ExecutionStartedEvent(-1, Raw(request.Input))
```
I think you can use `request.Input` directly here. No need to wrap it.
```suggestion
                            Event = new ExecutionStartedEvent(-1, request.Input)
```

davidmrdavid replied:
As per Jacob's suggestion, I'll use the TaskHubClient instead to avoid regressing on Distributed Tracing

davidmrdavid committed:
use TaskHubClient instead

davidmrdavid commented on src/WebJobs.Extensions.DurableTask/LocalGrpcListener.cs:
```diff
@@ -232,13 +246,13 @@ public async override Task<P.QueryEntitiesResponse> QueryEntities(P.QueryEntitie
                 EntityBackendQueries.EntityQueryResult result = await entityOrchestrationService.EntityBackendQueries!.QueryEntitiesAsync(
                     new EntityBackendQueries.EntityQuery()
                     {
-                         InstanceIdStartsWith = query.InstanceIdStartsWith,
-                         LastModifiedFrom = query.LastModifiedFrom?.ToDateTime(),
-                         LastModifiedTo = query.LastModifiedTo?.ToDateTime(),
-                         IncludeTransient = query.IncludeTransient,
-                         IncludeState = query.IncludeState,
-                         ContinuationToken = query.ContinuationToken,
-                         PageSize = query.PageSize,
+                        InstanceIdStartsWith = query.InstanceIdStartsWith,
+                        LastModifiedFrom = query.LastModifiedFrom?.ToDateTime(),
+                        LastModifiedTo = query.LastModifiedTo?.ToDateTime(),
+                        IncludeTransient = query.IncludeTransient,
+                        IncludeState = query.IncludeState,
+                        ContinuationToken = query.ContinuationToken,
+                        PageSize = query.PageSize,
```
ran a formatter and it fixed this.

jviau commented on src/WebJobs.Extensions.DurableTask/LocalGrpcListener.cs:
```diff
@@ -155,8 +155,22 @@ public async override Task<P.CreateInstanceResponse> StartInstance(P.CreateInsta
-                    string instanceId = await this.GetClient(context).StartNewAsync(
-                        request.Name, request.InstanceId, Raw(request.Input));
+                    string instanceId = request.InstanceId ?? Guid.NewGuid().ToString("N");
+                    TaskHubClient taskhubClient = new TaskHubClient(this.GetDurabilityProvider(context));
```
We may want to also provider `ILoggerFactory` to the ctor

jviau commented on src/WebJobs.Extensions.DurableTask/LocalGrpcListener.cs:
```diff
@@ -155,8 +155,22 @@ public async override Task<P.CreateInstanceResponse> StartInstance(P.CreateInsta
+                    TaskHubClient taskhubClient = new TaskHubClient(this.GetDurabilityProvider(context));
+
+                    // TODO: Ideally, we'd have a single method in the taskhubClient that can handle both scheduled and non-scheduled starts.
+                    // TODO: the type of `ScheduledStartTimestamp` is not nullable. Can we change it to `DateTime?` in the proto file?
```
I believe proto uses its own type: `Timestamp`, which is an object, so it can be `null` (whereas `DateTime` and `DateTimeOffset` are structs). So no, we cannot change it to `DateTime?`.

jviau commented on src/WebJobs.Extensions.DurableTask/LocalGrpcListener.cs:
```diff
@@ -155,8 +155,22 @@ public async override Task<P.CreateInstanceResponse> StartInstance(P.CreateInsta
+                    string instanceId = request.InstanceId ?? Guid.NewGuid().ToString("N");
+                    TaskHubClient taskhubClient = new TaskHubClient(this.GetDurabilityProvider(context));
+
+                    // TODO: Ideally, we'd have a single method in the taskhubClient that can handle both scheduled and non-scheduled starts.
```
Yeah, this was addressed in the durable-dotnet `DurableTaskClient` abstraction.

jviau commented on src/WebJobs.Extensions.DurableTask/LocalGrpcListener.cs:
```diff
@@ -155,8 +155,22 @@ public async override Task<P.CreateInstanceResponse> StartInstance(P.CreateInsta
+                        await taskhubClient.CreateOrchestrationInstanceAsync(request.Name, request.Version, instanceId, Raw(request.Input));
+                    }
+
+                    // TODO: should this not inclide the ExecutionId and other elements of the taskhubClient response?
```
We never included those in the proto. Something to consider for later.

jviau reviewed (approved):
LGTM. I wonder if we should also use `TaskHubClient` in other locations in this file? Probably for another PR/time.

davidmrdavid committed:
minor edit

developmentvegas commented:
Hi, Is there any update on this? Several users are stuck trying to use the startAt and we can't  ;(

davidmrdavid commented:
This is missing some tests still unfortunately. Looking to get back to this soon, but have some urgent competing threads at the moment. If anyone is willing to contribute some tests in the meantime (described in the PR description), I'd be happy to incorporate them.

tomseida commented:
We are approaching 11 months since the ticket was open presenting this defect. Any ETA when this will be available?

nikosdelis commented:
Is this missing anything else now that @cgillum's comments have been resolved?

nikosdelis commented:
Thanks a million @cgillum and @davidmrdavid for bringing this closer to its completion. Any idea about when we can expect this to be shipped?

tomseida commented:
Why has this pull request stalled? When can we expect these changes to be published? We are now at 13 months since this [defect was opened](https://github.com/microsoft/durabletask-dotnet/issues/256).

cgillum commented:
Sorry for the delay. David has moved to a different team and @andystaples and I should be able to take over this PR shortly. We're currently setting up some new end-to-end test infrastructure to make sure these kinds of regressions don't happen again. Fingers crossed that we can have this done and merged in the next couple weeks. 🤞

jharbieh-microsoft commented:
> Sorry for the delay. David has moved to a different team and @andystaples and I should be able to take over this PR shortly. We're currently setting up some new end-to-end test infrastructure to make sure these kinds of regressions don't happen again. Fingers crossed that we can have this done and merged in the next couple weeks. 🤞

Thank you @cgillum ... @tomseida is my customer and we were looking for direction on this.

andystaples committed:
Add scheduled start time E2E Test

cgillum commented on test/e2e/Apps/BasicDotNetIsolated/HelloCities.cs:
```diff
@@ -54,5 +54,27 @@ public static async Task<HttpResponseData> HttpStart(
+            [HttpTrigger(AuthorizationLevel.Anonymous, "get", "post")] HttpRequestData req,
+            [DurableClient] DurableTaskClient client,
+            FunctionContext executionContext,
+            DateTime ScheduledStartTime)
```
nit: method parameters should be camelCase: `scheduledStartTime`.

andystaples replied:
Done

cgillum commented on test/e2e/Tests/Helpers/DurableHelpers.cs:
```diff
@@ -0,0 +1,39 @@
+    }
+    internal static string GetRuntimeStatus(string statusQueryGetUri)
+    {
+        HttpClient client = new HttpClient();
```
We should avoid creating `HttpClient` objects and letting them get garbage collected since that can create a resource leak (customers run into this problem a lot). A better practice is to create a shared `static readonly HttpClient` that can be reused. They are safe for use by multiple threads concurrently, so you don't need to worry about synchronizing access to it.

andystaples replied:
Done

cgillum commented on test/e2e/Tests/Tests/HelloCitiesTest.cs:
```diff
@@ -19,17 +19,34 @@ public HttpEndToEndTests(FunctionAppFixture fixture, ITestOutputHelper testOutpu
+    [InlineData("HelloCities_HttpStart_Scheduled", HttpStatusCode.Accepted)]
+    public async Task ScheduledStartTests(string functionName, HttpStatusCode expectedStatusCode)
+    {
+        var scheduledStartDate = DateTime.Now + TimeSpan.FromSeconds(10);
```
Should this be `DateTime.UtcNow` instead of `DateTime.Now`?

andystaples replied:
Done

andystaples committed:
Test improvements
- Add a run for an orchestration scheduled in past
- Basic test case should check output
- Provide more info from StatusQueryGetUri response

andystaples committed:
Use UTC universally

dirkstrauss commented:
The absence of the `ScheduledStartTime` functionality in the .NET isolated process is currently impeding the development of a Durable Function for a client project. The resolution provided in this Pull Request is critical for our progress.
Could you please provide an estimated timeline for when this fix will be included in a stable release? This information is essential for our project planning and client communication.

tomseida commented:
@jharbieh-microsoft @cgillum, please provide an expected release date for the fix to this defect that was identified 14 months ago. It appears the momentum to get this released from a month ago is gone.

cgillum commented:
@bachuv or @nytian did we do an extension release recently that would have contained this fix? And is there a corresponding .NET isolated fix that picks it up? If not, let’s prioritize it this week if possible.

bachuv commented:
@cgillum Yes, the latest extension release contains the fix in this PR - https://github.com/Azure/azure-functions-durable-extension/releases/tag/v3.0.3, but we haven't released the corresponding .NET Isolated fix. We will prioritize this.

bachuv commented:
We discussed this offline and we will release the .NET Isolated packages with this fix next week.

jharbieh-microsoft commented:
Thank you team!

epDugas commented:
> We discussed this offline and we will release the .NET Isolated packages with this fix next week.

Fix available for .net isolated in https://www.nuget.org/packages/Microsoft.Azure.Functions.Worker.Extensions.DurableTask/1.2.3
